Add blackpill_f401ce

coderkalyan opened the pull request:
I just copied the blackpill_f411ce board definition for the newer blackpill_f401ce board, which is almost identical other than reduced clock and SRAM. Please let me know if I missed anything I'm supposed to add, this is my first PR to this project.

Also, added support for flashing over jlink for both blackpill models.

coderkalyan commented:
PS: I didn't quite understand from the contributing doc, and since the CI compliance is failing, could someone explain to me what the process is as far as allowed authors? Am I not eligible to submit a PR for this project or something?

lochej commented on boards/arm/blackpill_f401ce/blackpill_f401ce.dts:
```diff
@@ -0,0 +1,117 @@
+
+/dts-v1/;
+#include <st/f4/stm32f401Xe.dtsi>
+#include <st/f4/stm32f411c(c-e)ux-pinctrl.dtsi>
```
Thank you for your contribution 👍 
Please include the correct pinctrl definition. (even though there are not so many difference between f411ce and f401ce).
```suggestion
#include <st/f4/stm32f401c(d-e)ux-pinctrl.dtsi>
```

coderkalyan replied:
@lochej Thanks for reminding me, I was going to mention that in my PR. Could you please explain the syntax on line 9? I don't understand what (if any) regex/replace is being performed and what file it is referencing. When I simply replaced f411 to f401 west was unable to find the right file. In your suggestion, you said change the c to a d, why is that?

lochej replied:
We recently moved from pinmux.c to device tree pinmux definition using the pinctrl device. So users need to get used to it 😄 

Those includes are automatically generated in the **hal_stm32** repository.
Here is the path where you can find all the pinctrl.dtsi:
```
zephyrproject/modules/hal/stm32/dts/st/f4/
```

Then find the dtsi which matches your SOC. I do it manually most of the time when porting or adding boards.

lochej commented:
> 
> 
> PS: I didn't quite understand from the contributing doc, and since the CI compliance is failing, could someone explain to me what the process is as far as allowed authors? Am I not eligible to submit a PR for this project or something?

You have done a great job, actually, but you miss the ```Signed off``` statement at the end of your commit message.

For this you actually need to call ```git commit -s``` when creating a commit on your branch, then use the opened text editor to write your commit message.

If you want to check your latest commits for the code style warnings, use : 
```
./scripts/checkpatch.pl -g HEAD-1
```
you can check more than 1 commit for the their code style validation by replacing the ```1``` by ```5``` to check the last 5 commits.

erwango commented on boards/arm/blackpill_f401ce/blackpill_f401ce.dts:
```diff
@@ -0,0 +1,117 @@
+
+/ {
+	model = "WeAct Studio Black Pill V3.0 Board";
+	compatible = "weact,blackpill-f401ce","st,stm32f401";
```
Please remove the soc compatible:
```
	compatible = "weact,blackpill-f401ce";
```

erwango commented on boards/arm/blackpill_f401ce/doc/index.rst:
```diff
@@ -0,0 +1,179 @@
+=========
+
+The board can be debugged by installing the included 100 mil (0.1 inch) header,
+andattaching an SWD debugger to the 3V3 (3.3V), GND, SCK, and DIO
```
`and attaching`

gmarull commented on boards/arm/blackpill_f401ce/support/openocd.cfg:
```diff
@@ -0,0 +1,12 @@
+source [find board/stm32f4discovery.cfg]
```
Even though it likely works, the cfg is for another board, so unless this board is supported by openocd better use something like

```
source [find interface/stlink.cfg]

transport select hla_swd

source [find target/stm32f4x.cfg]

reset_config srst_only
```

gmarull reviewed (commented):
If only the SoC is the only part that is different you may consider adding the 2 variations in a single board folder. @erwango is that possible?

erwango commented:
> If only the SoC is the only part that is different you may consider adding the 2 variations in a single board folder. @erwango is that possible?

This kind if things is doable for boards revision variants, though it does not include the SoC change. Having a variation on SoC only is actually also the case for a lot of boards (nucleos for instance, and I guess this is the case for other Soc and boards vendors as well, olimex? particle ? ...). Maybe there is some room for improvement in that direction but I think we should study the impacts first, in matter of docs CI, ...

coderkalyan commented:
@lochej @gmarull @erwango Thank you all for your reviews! I have fixed all of the issues (I think) including the typo, openocd config, compatible line in the dts, and the dts include. I also signed off my commits, so hopefully that works fine now :smile:  I also took the liberty to apply these fixes to the blackpill_f411ce board as well, as it seems to have been merged to master without a detailed review.

Hopefully this is fine for merging now!

jukkar commented:
> If only the SoC is the only part that is different you may consider adding the 2 variations in a single board folder.

Couple of months ago I tried to put two board variants to a single board folder. This did not work that well because the script, that generates the HTML pages, got confused and only produced documentation link to one board instead of two. I did not notice any other issue at that time.

lochej reviewed (changes requested):
If you want to pass BuildKite, please add your board to the **test/drivers/adc/adc_api/src/test_adc.c**:

```c
#elif defined(CONFIG_BOARD_NUCLEO_L476RG) || \
	defined(CONFIG_BOARD_BLACKPILL_F411CE) || \
	defined(CONFIG_BOARD_BLACKPILL_F401CE) || \    <---- Add this
	defined(CONFIG_BOARD_NUCLEO_L4R5ZI)
#define ADC_DEVICE_NAME         DT_LABEL(DT_INST(0, st_stm32_adc))
#define ADC_RESOLUTION		10
#define ADC_GAIN		ADC_GAIN_1
#define ADC_REFERENCE		ADC_REF_INTERNAL
#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
#define ADC_1ST_CHANNEL_ID	1
```

lochej commented on boards/arm/blackpill_f401ce/blackpill_f401ce.dts:
```diff
@@ -0,0 +1,117 @@
+
+	pwm4: pwm {
+		status = "okay";
+		pinctrl-0 = <&tim4_ch1_pwm_pb6 &tim4_ch2_pwm_pb7>;
```
```suggestion
		pinctrl-0 = <&tim4_ch1_pb6 &tim4_ch2_pb7>;
```
Please rebase on latest master and change the timer pins definitions to the following.
Some update to hal_stm32 changed phandles names.
See #29729 for example.

That's why your buildkite fails.

coderkalyan commented:
@lochej Thanks! I just pushed, I'll check that the CI builds successfully this time but it should be good now.

A couple of questions, regarding this board and how it relates to the device tree system (I'm really new to zephyr and still trying to figure out the peripheral APIs):
1) The board's DTS file (that I primarily copied from the blackpill_f411 definition) only enables timer4. What is the procedure for enabling other timers? Should I enable all of them in this file and update the PR? Or is there another project-level DTS file I should use to enable timers (or other peripherals that are disabled, if I come across any)?
2) What's the best way to get general help if I'm having trouble with something? Is there a forum or public chat somewhere? All I found was a link to a Slack but that seemed reserved for certain emails. Should I just be filing issues on the github repository?

coderkalyan commented:
Update to my previous question: I was able to add an overlay file to my project that enables the timers and PWM devices. Is this something that should remain in the overlay, or should it be enabled in the main board DTS file created in this PR?

lochej commented:
> Update to my previous question: I was able to add an overlay file to my project that enables the timers and PWM devices. Is this something that should remain in the overlay, or should it be enabled in the main board DTS file created in this PR?

Keeping it in the overlay is good enough !
What should be in the main board dts are required peripheral declarations for the test framework and samples. But I don't know if there are rules for that.

Then in your app, you can enable as many peripherals as you want using the overlay.

erwango reviewed (changes requested):
Please squash the thrid commit into the first one (we don't merge "fix commit")

coderkalyan commented:
> Please squash the thrid commit into the first one (we don't merge "fix commit")

@erwango Do you mean merge the second commit onto the first? I only have 2 commits, one for adding the f401 board, the other for updating the f411 board.

erwango commented:
> @erwango Do you mean merge the second commit onto the first? I only have 2 commits, one for adding the f401 board, the other for updating the f411 board.

My bad. Indeed please merge the 2 commits.
Also, in the other hand, modification to test/... should be done in a dedicated commit.

coderkalyan commented:
@erwango I merged the two commits... buildkite is still pending but it should be fine :smile:

lochej commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -176,6 +176,7 @@
 
 #elif defined(CONFIG_BOARD_NUCLEO_L476RG) || \
 	defined(CONFIG_BOARD_BLACKPILL_F411CE) || \
+	defined(CONFIG_BOARD_BLACKPILL_F401CE) || \
 	defined(CONFIG_BOARD_NUCLEO_L4R5ZI)
```
Please make a dedicated commit for this addition please :)

coderkalyan replied:
Ok, will do.

gmarull commented on boards/arm/blackpill_f411ce/blackpill_f411ce.dts:
```diff
@@ -10,7 +10,7 @@
 
 / {
 	model = "WeAct Studio Black Pill V2.0 Board";
-	compatible = "weact,blackpill-f411ce","st,stm32f411";
```
changes to `f411ce` board should go into a separate commit.

coderkalyan replied:
I thought so too, which was originally what I had, but @erwango suggested that I combine the 2 commits, so I did. If you want I can split it back up.

erwango replied:
Did I ?
Ho got it! I thought this second commit was a bunch of fixes on the initial commit (as this often happened recently).
My bad, I was too fast. Yes 2 commits are required as these are different boards. Sorry @coderkalyan for the back and forth changes

coderkalyan replied:
No worries, it happens :smiley: I'll split up the commits again soon.

gmarull reviewed (changes requested):
LGTM, however, I think f411 changes should go into a separate commit

coderkalyan committed:
boards: arm: add blackpill_f401ce board definition

Duplicates the existing blackpill_f411ce board definition for the
blackpill_f401ce board (WeAct Blackpill V3.0), which has identical specs
except for 84MHz clock (instead of 100Mhz) and 96K SRAM (instead of
128K).

Signed-off-by: Kalyan Sriram <coder.kalyan@gmail.com>

coderkalyan committed:
boards: arm: update blackpill_f411ce definition

Discussion while adding the blackpill_f401ce board definition brought
up a few issues with the blackpill_f411ce board definition, which are
addressed in this commit:
* doc/index.rst had a typo
* support/openocd.cfg was incorrectly using the f4discovery profile,
  which was fixed by using a custom profile for the SoC instead
* removed SoC from blackpill_f411ce.dts compatible line
* added support for jlink flash option

Signed-off-by: Kalyan Sriram <coder.kalyan@gmail.com>

coderkalyan committed:
boards: arm: enable adc test for blackpill_f401ce

Adds the blackpill_f401ce board configuration to the
adc driver test.

Signed-off-by: Kalyan Sriram <coder.kalyan@gmail.com>

coderkalyan commented:
@gmarull @erwango @lochej I pulled the f411 changes back into their own commit, and pulled the adc test addition into its own commit as well.

Thanks a lot for your repeated reviews and assistance! :smiley: Hopefully this is now good to go!
